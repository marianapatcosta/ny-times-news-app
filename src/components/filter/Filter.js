import { defineComponent } from "vue";

export default defineComponent({
  name: "Filter",
  emits: ["select-option"],
  props: {
    name: {
      type: String,
      required: false,
    },
    options: {
      type: Array,
      required: true,
    },
    placeholder: {
      type: String,
      required: false,
    },
    disabled: {
      type: Boolean,
      required: false,
    },
  },
  data() {
    return {
      selectedOption: "",
      showOptions: false,
      searchKeyword: "",
    };
  },
  computed: {
    filteredOptions() {
      return this.options.filter((option) =>
        option.toLowerCase().includes(this.searchKeyword.toLowerCase())
      );
    },
  },
  methods: {
    toggleShowOptions() {
      this.showOptions = !this.showOptions;
    },
    handleOptionSelection(option) {
      this.selectedOption = option === "All" ? "" : option;
      this.showOptions = false;
      this.searchKeyword = "";
      this.$emit("select-option", this.selectedOption);
    },
  },
});
