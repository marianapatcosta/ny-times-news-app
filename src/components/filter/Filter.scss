.filter {
  position: relative;
  width: 10rem;

  &__button {
    background-color: $color-secondary;
    box-shadow: 0 0.05rem 0.3rem rgba($color-black, 0.5);
    color: $color-primary;
    cursor: pointer;
    display: block;
    padding: 0.5rem 1rem;
    width: 100%;
  }

  &__search {
    background-color: $color-primary;
    color: $color-secondary;
    position: sticky;
    top: 0;
    
    input {
      margin-left: 0.125rem;
      margin-top: 0.165rem;
      padding: 0.2rem;
      width: calc(100% - 0.25rem);
    }

    & input::placeholder {
      color: $color-secondary;
      font-style: italic;
    }
  }

  &__content {
    list-style: none;
    max-height: 8rem;
    position: absolute;
    overflow: auto;
    overflow-x: hidden;
    background-color: $color-primary;
    width: 100%;
    z-index: 1;
  }

  &__option {
    cursor: pointer;
    padding: 0.2rem;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    user-select: none;

    &--active {
      background-color: $color-secondary;
      color: $color-primary;
    }

    &:hover {
      background-color: $color-secondary;
      color: $color-primary;
    }
  }
}
