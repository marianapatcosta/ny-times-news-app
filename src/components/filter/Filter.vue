<template>
  <div class="filter" v-if="options">
    <button
      class="filter__button"
      aria-label="Click to check filter options"
      @click="toggleShowOptions()"
      @keydown.enter="toggleShowOptions()"
    >
      {{ selectedOption || "Filter by" }}
    </button>
    <ul class="filter__content" v-show="showOptions">
      <li class="filter__search">
        <input
          :disabled="disabled"
          :placeholder="placeholder || 'Search...'"
          :aria-label="placeholder || 'Search...'"
          v-model="searchKeyword"
        />
      </li>
      <li
        class="filter__option"
        :class="{ 'filter__option--active': option === selectedOption }"
        v-for="option in filteredOptions"
        :key="`option-${option}`"
        role="button"
        tabIndex="0"
        :aria-label="`Click to filter ${option}`"
        @click="handleOptionSelection(option)"
        @keydown.enter="handleOptionSelection(option)"
      >
        {{ option || "-" }}
      </li>
    </ul>
  </div>
</template>

<script src="./Filter.js"></script>

<style scoped lang="scss" src="./Filter.scss"></style>
