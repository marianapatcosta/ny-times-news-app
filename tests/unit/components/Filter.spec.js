import { mount } from "@vue/test-utils";
import Filter from "@/components/filter/Filter.vue";

describe("Filter.vue", () => {
  const options = ["All", "Option 1", "Option 2"];
  let wrapper;
  beforeEach(() => {
    wrapper = mount(Filter, {
      props: { options },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });
  it("should render without errors and match snapshot renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should not show filter option when rendered", async () => {
    const filterOptions = wrapper.find("ul");
    expect(filterOptions.isVisible()).toBe(false);
  });

  it("should not show filter option when rendered", async () => {
    const filterOptions = wrapper.find("ul");
    const button = wrapper.find(".filter__button");
    await button.trigger("click");
    expect(filterOptions.isVisible()).toBe(true);
  });

  it("should render 'Filter by' when no option is selected", () => {
    const button = wrapper.find(".filter__button");
    expect(button.text()).toMatch("Filter by");
  });

  it("should render option when a filter option is selected", async () => {
    const filterOptions = wrapper.find("ul").findAll("li");
    const button = wrapper.find(".filter__button");
    await button.trigger("click");
    await filterOptions.at(2).trigger("click");
    expect(button.text()).toMatch("Option 1");
  });

  it("should render 'Filter by' when 'All' option is selected", async () => {
    const filterOptions = wrapper.find("ul").findAll("li");
    const button = wrapper.find(".filter__button");
    await button.trigger("click");
    await filterOptions.at(1).trigger("click");
    expect(button.text()).toMatch("Filter by");
  });

  it("should render 'Search by' in search box when props.placeholder is not passed", () => {
    const search = wrapper.find(".filter__search input");
    expect(search.attributes("placeholder")).toMatch("Search...");
  });

  it("should render props.placeholder in search box", () => {
    const placeholder = "Search by";
    const wrapper = mount(Filter, {
      props: { options, placeholder },
    });
    const search = wrapper.find(".filter__search input");
    expect(search.attributes("placeholder")).toMatch(placeholder);
  });

  it("should emit 'select-option' event when a filter option is selected", async () => {
    const filterOptions = wrapper.find("ul").findAll("li");
    const button = wrapper.find(".filter__button");
    await button.trigger("click");
    await filterOptions.at(2).trigger("click");
    expect(wrapper.emitted()).toHaveProperty("select-option");
    expect(wrapper.emitted()["select-option"][0]).toEqual(["Option 1"]);
  });
});
