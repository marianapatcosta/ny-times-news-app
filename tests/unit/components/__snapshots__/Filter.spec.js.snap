// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter.vue should render without errors and match snapshot renders correctly 1`] = `
<div
  class="filter"
>
  <button
    aria-label="Click to check filter options"
    class="filter__button"
  >
    Filter by
  </button>
  <ul
    class="filter__content"
    style="display: none;"
  >
    <li
      class="filter__search"
    >
      <input
        aria-label="Search..."
        placeholder="Search..."
      />
    </li>
    
    <li
      aria-label="Click to filter All"
      class="filter__option"
      role="button"
      tabindex="0"
    >
      All
    </li>
    <li
      aria-label="Click to filter Option 1"
      class="filter__option"
      role="button"
      tabindex="0"
    >
      Option 1
    </li>
    <li
      aria-label="Click to filter Option 2"
      class="filter__option"
      role="button"
      tabindex="0"
    >
      Option 2
    </li>
    
  </ul>
</div>
`;
